solver/src/http_solver.rs

```diff
@@ -0,0 +1,270 @@
+}
+
+impl SolverConfig {
+    fn add_to_query(&self, url: &mut Url) {
+        url.query_pairs_mut()
+            .append_pair(
+                "max_nr_exec_orders",
+                self.max_nr_exec_orders.to_string().as_str(),
+            )
+            .append_pair("time_limit", self.time_limit.to_string().as_str());
+    }
+}
+
+pub struct HttpSolver {
+    base: Url,
+    client: Client,
+    api_key: Option<String>,
+    config: SolverConfig,
+    uniswap: UniswapV2Factory,
+}
+
+impl HttpSolver {
+    pub fn new(
+        base: Url,
+        api_key: Option<String>,
+        config: SolverConfig,
+        uniswap: UniswapV2Factory,
+    ) -> Self {
+        // Unwrap because we cannot handle client creation failing.
+        let client = Client::builder().build().unwrap();
+        Self {
+            base,
+            client,
+            api_key,
+            config,
+            uniswap,
+        }
+    }
+
+    // Solver api requires specifying token as strings. We use the address as a string for now.
+    // Later we could use a more meaningful name like the token symbol but we have to ensure
+    // uniqueness.
+    fn token_to_string(&self, token: &H160) -> String {
+        // Token names must start with a letter.
+        format!("t{:x}", token)
+    }
+
+    fn tokens(&self, orders: &[Order]) -> HashMap<String, TokenInfoModel> {
+        orders
+            .iter()
+            .flat_map(|order| {
+                let order = order.order_creation;
+                std::iter::once(order.sell_token).chain(std::iter::once(order.buy_token))
+            })
+            .collect::<HashSet<_>>()
+            .into_iter()
+            .map(|token| {
+                // TODO: gather real decimals and store them in a cache
+                let token_model = TokenInfoModel { decimals: 18 };
+                (self.token_to_string(&token), token_model)
+            })
+            .collect()
+    }
+
+    fn orders(&self, orders: &[Order]) -> HashMap<String, OrderModel> {
+        orders
+            .iter()
+            .enumerate()
+            .map(|(index, order)| {
+                let order = order.order_creation;
+                let order = OrderModel {
+                    sell_token: self.token_to_string(&order.sell_token),
+                    buy_token: self.token_to_string(&order.buy_token),
+                    sell_amount: order.sell_amount,
+                    buy_amount: order.buy_amount,
+                    allow_partial_fill: order.partially_fillable,
+                    is_sell_order: matches!(order.kind, OrderKind::Sell),
+                };
+                (index.to_string(), order)
+            })
+            .collect()
+    }
+
+    async fn uniswaps(&self, orders: &[Order]) -> Result<HashMap<String, UniswapModel>> {
+        // TODO: use a cache
+        let mut uniswaps = HashMap::new();
+        for order in orders {
+            let pair = order.order_creation.token_pair().expect("invalid order");
```

Comment:
I'm having doubts about this. This code is getting the uniswap pool for every token pair in the orders. But what if there is a third pool for which there is no order which nonetheless connects the token better. We probably want the direct pools AND every individual token with ETH since that is usually the biggest pool. Then the question is whether the solver is aware of this too.

Reply:
>We probably want the direct pools AND every individual token with ETH since that is usually the biggest pool.
Yeah agreed.

>Then the question is whether the solver is aware of this too.
Sure, the solvers will be able to trade "rings" through many pools